File: frontend/mento/src/types/product.ts
// API 상품 정보 타입 (백엔드 응답 형식)
export interface ApiProduct {
  productId: number
  oliveyoungGoodsNo: string
  brandId: number
  brandName: string
  categoryMedium: string
  categorySmall: string
  name: string
  volume: string
  description?: string
  ingredients?: string
  price: number
  imageUrl: string
  productUrl: string
  skinTypes?: string  // JSON 배열 문자열
  relatedConditions?: string  // JSON 배열 문자열
  benefits?: string  // JSON 배열 문자열
  defaultUsageDays?: number
  createdAt: string
  updatedAt?: string
}

// 페이지네이션 응답 타입
export interface PaginatedResponse<T> {
  content: T[]
  page: number
  size: number
  totalElements: number
  totalPages: number
  first: boolean
  last: boolean
}

// 공통 API 응답 래퍼 타입
export interface ApiResponse<T> {
  success: boolean
  data: T
  error?: {
    code: string
    message: string
  }
  timestamp: string
}

// 상품 목록 아이템 타입
export interface ProductListItem {
  productId: number
  name: string
  brandName: string
  categoryMedium: string
  imageUrl: string
}

// 상품 목록 조회 파라미터
export interface ProductListParams {
  page?: number
  size?: number
}

// 정렬 옵션 타입
export type SortKey = "price" | "name" | "created_at"
export type SortOrder = "asc" | "desc"

// 카테고리 타입 (실제 카테고리 값은 백엔드 응답에 따라 수정 필요)
export type CategoryMedium = string
export type CategorySmall = string

File: backend/src/main/java/com/mento/domain/product/entity/Product.java
package com.mento.domain.product.entity;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import com.mento.common.entity.BaseEntity;
import com.mento.common.error.ErrorCode;
import com.mento.domain.brand.entity.Brand;
import com.mento.domain.item.entity.UserItem;
import com.mento.domain.product.exception.ProductException;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PROTECTED)
@Table(name = "products")
public class Product extends BaseEntity {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "product_id")
	private Long id;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "brand_id", nullable = false)
	private Brand brand;

	@Builder.Default
	@OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
	private List<UserItem> userItems = new ArrayList<>();

	@Column(name = "oliveyoung_goods_no", nullable = false, length = 50)
	private String oliveyoungGoodsNo;

	@Column(name = "category_medium", length = 100)
	private String categoryMedium;

	@Column(name = "category_small", length = 100)
	private String categorySmall;

	@Column(nullable = false, length = 500)
	private String name;

	@Column(length = 255)
	private String volume;

	@Column(length = 1000)
	private String description;

	@Column(columnDefinition = "TEXT")
	private String ingredients;

	@Column(nullable = false)
	@Builder.Default
	private Integer price = 0;

	@Column(name = "image_url", length = 500)
	private String imageUrl;

	@Column(name = "product_url", length = 1000)
	private String productUrl;

	@JdbcTypeCode(SqlTypes.JSON)
	@Column(name = "skin_types", columnDefinition = "json")
	private List<String> skinTypes;

	@JdbcTypeCode(SqlTypes.JSON)
	@Column(name = "related_conditions", columnDefinition = "json")
	private List<String> relatedConditions;

	@JdbcTypeCode(SqlTypes.JSON)
	@Column(name = "benefits", columnDefinition = "json")
	private List<String> benefits;

	@Column(name = "default_usage_days", nullable = false)
	@Builder.Default
	private Integer defaultUsageDays = 90;

	@Column(name = "deleted_at")
	private LocalDateTime deletedAt;

	public void assignUserItem(final UserItem userItem) {
		if (userItem == null) {
			throw new ProductException(ErrorCode.MISSING_USER_ITEM);
		}
		userItems.add(userItem);
		if (userItem.getProduct() != this) {
			userItem.assignProduct(this);
		}
	}
}
